### ISSUE TITLE
Crash when running sqlfluff fix on query that contains JOIN with subquery

### ISSUE THREAD

**Issue by kien-truong** [sqlfluff/sqlfluff#6456]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Running `sqlfluff fix` on this file will crash it with an assertion error

```
WITH x AS (
  WITH y AS (
    SELECT 1 AS a
  )
  SELECT * FROM y
  CROSS JOIN (SELECT 1 AS a)
)
SELECT * FROM x
```

```
Unable to lint test.sql due to an internal error. Please report this as an issue with your query's contents and stacktrace below!
To hide this warning, add the failing file to .sqlfluffignore
Traceback (most recent call last):
  File "/home/kien_truong/.local/pipx/venvs/sqlfluff/lib/python3.11/site-packages/sqlfluff/core/rules/base.py", line 511, in crawl
    res = self._eval(context=context)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kien_truong/.local/pipx/venvs/sqlfluff/lib/python3.11/site-packages/sqlfluff/rules/capitalisation/CP02.py", line 114, in _eval
    return super()._eval(context=context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kien_truong/.local/pipx/venvs/sqlfluff/lib/python3.11/site-packages/sqlfluff/rules/capitalisation/CP01.py", line 91, in _eval
    return [self._handle_segment(context.segment, context)]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kien_truong/.local/pipx/venvs/sqlfluff/lib/python3.11/site-packages/sqlfluff/rules/capitalisation/CP01.py", line 129, in _handle_segment
    if segment.is_templated and ignore_templated_areas:
       ^^^^^^^^^^^^^^^^^^^^
  File "/home/kien_truong/.local/apps/mambaforge/envs/pipx/lib/python3.11/functools.py", line 1001, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/kien_truong/.local/pipx/venvs/sqlfluff/lib/python3.11/site-packages/sqlfluff/core/parser/segments/base.py", line 421, in is_templated
    assert self.pos_marker
AssertionError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kien_truong/.local/pipx/venvs/sqlfluff/lib/python3.11/site-packages/sqlfluff/core/linter/runner.py", line 115, in run
    yield partial()
          ^^^^^^^^^
  File "/home/kien_truong/.local/pipx/venvs/sqlfluff/lib/python3.11/site-packages/sqlfluff/core/linter/linter.py", line 789, in lint_rendered
    return cls.lint_parsed(
           ^^^^^^^^^^^^^^^^
  File "/home/kien_truong/.local/pipx/venvs/sqlfluff/lib/python3.11/site-packages/sqlfluff/core/linter/linter.py", line 649, in lint_parsed
    ) = cls.lint_fix_parsed(
        ^^^^^^^^^^^^^^^^^^^^
  File "/home/kien_truong/.local/pipx/venvs/sqlfluff/lib/python3.11/site-packages/sqlfluff/core/linter/linter.py", line 469, in lint_fix_parsed
    linting_errors, _, fixes, _ = crawler.crawl(
                                  ^^^^^^^^^^^^^^
  File "/home/kien_truong/.local/pipx/venvs/sqlfluff/lib/python3.11/site-packages/sqlfluff/core/rules/base.py", line 527, in crawl
    assert context.segment.pos_marker
AssertionError
```

### Expected Behaviour

`sqlfluff fix` do not crash

### Observed Behaviour

`sqlfluff fix` crash with an assertion error, however the return code is 0.

### How to reproduce

Running `sqlfluff fix` with the provided file.

### Dialect

`bigquery`

### Version

```
sqlfluff: version 3.2.5
python: version 3.11
```

### Configuration

```
❯ cat .sqlfluff
[sqlfluff]
dialect=bigquery
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by kien-truong** [comment]
Running `sqlfluff format` to reformat the file before running `sqlfluff fix` will avoid the crash.
